### ISSUE TITLE
Support login with verification code

### ISSUE THREAD

**Issue by ldzhjn** [FreshSupaSulley/Auto-2FA#11]
Can we support login with verification code?
Now Duochrome is throwing below exception
400 - Bad Request
{\"code\":40014,\"message\":\"Please enter the verification code to respond to this login.\",\"stat\":\"FAIL\"}"

The workflow is like this:
Duo login page will pop out with four or six digits verification code.
Enter these these digits in the Duochrome and press login.

**Comment by FreshSupaSulley** [comment]
Sorry I'm a bit confused. Is this a feature request, or is the new update breaking the functionality of older versions and is throwing that error?

Are you referring to functionality like [this](https://guide.duo.com/android#verified-push)? I would be down to add this feature but I don't have a way to test this workflow and put it into the extension. I'm a student and my organization uses push requests, not codes to approve logins so I can't study the HTTP traffic.

**Comment by FreshSupaSulley** [comment]
That would be great.

**Comment by FreshSupaSulley** [comment]
@ldzhjn I used Android Studio to setup an emulator and use HTTP Toolkit to sniff the packets. When the activation packet gets sent through, it attaches a "jailbroken: true" property to the request. You can use HTTP Toolkit to intercept this traffic, change it to false, and send it. But before you do any of that, there's the issue of SSL pinning. I used a Frida SSL unpinning script to view the HTTPS traffic.

This is a lot of work. Is there a way I can get an account for your organization and I can try reading the traffic instead?

**Comment by FreshSupaSulley** [comment]
Strange. Try different Frida scripts online? There should be a lot of SSL unpinning ones available.

**Comment by FreshSupaSulley** [comment]
For anyone else wanting this feature, I'll need either the OP or someone else to provide HTTP request logs that show the entire flow from transactions to sending the verification code to Duo

**Comment by cyber-sec0** [comment]
Requested code
https://api-d3b66583.duosecurity.com/frame/v4/auth/prompt/data?post_auth_action=OIDC_EXIT&browser_features=%7B%22touch_supported%22%3Afalse%2C%22platform_authenticator_status%22%3A%22unavailable%22%2C%22webauthn_supported%22%3Atrue%7D&sid=frameless-a3ce9bbf-5d91-4145-a69e-1113a2175349

Entered code
https://api-d3b66583.duosecurity.com/frame/v4/prompt
`  "body": "passcode=563427&device=null&factor=Passcode&postAuthDestination=OIDC_EXIT&browser_features=%7B%22touch_supported%22%3Afalse%2C%22platform_authenticator_status%22%3A%22unavailable%22%2C%22webauthn_supported%22%3Atrue%7D&sid=frameless-a3ce9bbf-5d91-4145-a69e-1113a2175349"`

https://api-d3b66583.duosecurity.com/frame/v4/status
`  "body": "txid=843620e6-db55-4a5b-a5d8-1dba1407aa6e&sid=frameless-a3ce9bbf-5d91-4145-a69e-1113a2175349"`

**Comment by kwshi** [comment]
@raxod502 and I did some reverse-engineering and seem to have figured out that the way to submit an approval request is to pass the code in a `step_up_code` query parameter:

https://github.com/FreshSupaSulley/Auto-2FA/blob/75cff6c1de8cc8833b5860ed593c144de0460472/firefox/worker.js#L190

e.g., replace `{ answer: "approve" }` with `{ answer: "approve", step_up_code: "1234" }`

It worked for me; perhaps someone else can try and see if it works for them as well?

**Comment by FreshSupaSulley** [comment]
Thanks everybody. Please bear with me as I'm still a bit lost on the flow. I also can't test this on my end because my org doesn't support verified Duo push. Might need to create a free trial admin account...

TLDR: I need to know if there's a flag in the transactions endpoint that indicates if it's a verified Duo push so I don't show the user a box to input their `step_up_code` when it's not actually a verified push. I imagine the flag exists (either in the transactions endpoint or elsewhere) as the mobile app shows a popup box when needed.

Not exactly sure, but I think @cyber-sec0 confusing TOTP and Duo verified push (that's #18, and it's in the works!)

@kwshi sort of a big ask, but could you either A) let me know how to differentiate verified pushes from standard ones or B) scrub the relevant HTTP requests (to not reveal anything sensitive) and post them here? I bet you can get away with just letting me know which endpoint has the flag and what it's called.

**Comment by cyber-sec0** [comment]
I'm not getting confused. I know the difference between both.

I don't think that the extension needs a duo code verification, it already works well and perfectly for Duo. I wouldn't change a single thing regarding duo for now.

What I want implemented is a totally different non duo related OTP pin code generator. That any extension (that doesn't support duo) has.

-------- Original Message --------

**Comment by kwshi** [comment]
@FreshSupaSulley here is a diff between `/push/v2/device/transactions` responses for basic push and verified push:

```diff
--- basic.json	2025-05-11 13:15:32.264462879 -0700
+++ verified.json	2025-05-11 13:15:47.140818485 -0700
@@ -1,32 +1,33 @@
 {
 	"response": {
 		"current_time": 1746993771,
 		"transactions": [
 			{
 				"access_device_ble_status": "unknown",
 				"akey": "BBBBBBBBBBBBBBBBBBBB",
 				"attributes": [
 					[["Organization", "XXXX XXXX"], ["Integration", "XXXX YYYY ZZZZ"]],
 					[
 						["Username", "foobar"],
 						["IP Address", "000.000.000.000"],
 						["Location", "Mars, Solar System"],
 						["Time", 1746994125.3203323]
 					]
 				],
 				"block_biometric_pin_fallback": false,
 				"expiration": 1746994184,
 				"is_proximity_push": false,
 				"is_strict_proximity_push": false,
 				"phone_notification_type": "visible_notification",
-				"require_lock": false,
+				"require_lock": true,
 				"require_touch_id": false,
+				"step_up_code_info": { "num_digits": 4 },
 				"summary": "XXXX YYYY ZZZZ",
 				"txid": "aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa",
 				"type": "Login",
 				"urgid": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
 			}
 		]
 	},
 	"stat": "OK"
 }
```

Does this help?

**Comment by FreshSupaSulley** [comment]
Yes, thank you. I suppose the extension would simply detect the presence of those JSON properties, and prompt the user to enter it themselves? (later it could pass the step up code automatically for zero click logins)

**Comment by kwshi** [comment]
Another idea is to scrape the current web page HTML to try to find the code (in the same spirit as the QR code detection for the activation flow)-- would that be something doable, do you think?

**Comment by FreshSupaSulley** [comment]
Yeah it would likely use the same message passing method as the QR code

**Comment by chenxiaoqino** [comment]
I'd be happy to help test the feature when there's an alpha version.

Also, here's the HTML from the page displaying verification code, hope it helps the auto parsing.

```
<div><div class="row display-flex "><h1 id="header-text" class="" tabindex="-1">Enter code in Duo Mobile</h1></div><div class="row display-flex instruction-text"><p>Verify it's you by entering this verification code in the Duo Mobile app...</p></div><div class="row display-flex align-flex-justify-content-center verification-code">321</div><div class="row display-flex align-flex-justify-content-center size-font-xsmall size-line-height-small size-margin-bottom-small"><span aria-hidden="true" class="phone-name">Sent to "Android" (•••-•••-1234)</span><span class="screen-reader-only-text">Sent to phone number ending in 2846</span></div><div class="row display-flex push-icon-container align-flex-justify-content-center align-flex-items-center"><span class="device-icon"><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="https://vecta.io/nano" width="90" height="84"><defs><linearGradient x1="50%" y1="0%" x2="50%" y2="18.3096493%" id="A"><stop stop-color="#fff" stop-opacity="0" offset="0%"></stop><stop stop-color="#fff" stop-opacity=".085" offset="20.5556163%"></stop><stop stop-color="#fff" offset="100%"></stop></linearGradient></defs><g fill="none" fill-rule="evenodd"><rect stroke="#c4ced4" stroke-width="3" fill="#e8eff3" x="1.5" y="1.5" width="87" height="99" rx="8"></rect><g fill="#dae5eb"><rect x="48" y="15" width="14" height="14" rx="4"></rect><rect x="8" y="35" width="14" height="14" rx="4"></rect><rect x="28" y="35" width="14" height="14" rx="4"></rect><rect x="68" y="35" width="14" height="14" rx="4"></rect><rect x="8" y="15" width="14" height="14" rx="4"></rect><rect x="28" y="15" width="14" height="14" rx="4"></rect><rect x="68" y="15" width="14" height="14" rx="4"></rect><rect x="48" y="55" width="14" height="14" rx="4"></rect><rect x="8" y="55" width="14" height="14" rx="4"></rect><rect x="28" y="55" width="14" height="14" rx="4"></rect><rect x="68" y="55" width="14" height="14" rx="4"></rect></g><rect fill="#68bb4e" x="45.789" y="32.368" width="18.667" height="18.667" rx="4"></rect><g fill="#fff"><path d="M48.456 43.918H50.6c1.147 0 2.084-.934 2.14-2.107h-4.283v2.107zM50.6 39.48h-2.143v2.107h4.283c-.056-1.174-.993-2.107-2.14-2.107m8.998 0c-1.147 0-2.084.934-2.14 2.107h4.28c-.056-1.174-.993-2.107-2.14-2.107m0 4.438c1.147 0 2.084-.934 2.14-2.107h-4.28c.056 1.174.993 2.107 2.14 2.107"></path><path d="M52.955 41.7c0 1.188.902 2.158 2.035 2.215V39.48h-2.035v2.22zm2.25-2.22v4.436h2.035V39.48z"></path><rect x="5.526" y="6" width="78.947" height="16" rx="2"></rect></g><path d="M19.737 11.053h18.947M72.63 11h6.316m-59.21 4.8H57.63" stroke="#dae5eb" stroke-width="2" stroke-linecap="round"></path><path d="M8.684 18.158v-3.454h7.4c-.1 1.9-1.722 3.454-3.694 3.454H8.684z" fill="#a5d692"></path><path d="M8.684 10.263v3.454h7.4c-.1-1.885-1.676-3.454-3.648-3.454H8.684z" fill="#6bbf4e"></path><path fill="url(#A)" d="M0 53h90v64H0z"></path></g></svg></span><span class="loader-icon-container"><span class="loader-icon bounce-in"><div class="loader loader--accessible loader--size-large loader--push-request" role="presentation"></div></span></span></div><div class="row display-flex align-flex-justify-content-center"><p class="loader-text">Waiting for approval...</p></div><div class="row display-flex other-options-link align-flex-justify-content-center size-margin-bottom-large size-margin-top-small"><a class="button--link " href="/frame/v4/auth/all_methods?sid=frameless-30bfc967-12e4-4eba-bcb5-08ac987e9149">Other options</a></div></div>
```

**Comment by kwshi** [comment]
Same for me-- the code is stored in a div with class `verification-code`.

**Comment by FreshSupaSulley** [comment]
Feel free to pull the head of the repo and load the chrome folder as an unpacked extension and stress test it to kingdom come (Firefox isn't updated yet, I want to make sure chrome works first). There's functionality for zero-clicks too, so be sure to give that a try. Very decent chance it will break. If it does, drop the logs / network requests.

Thanks for your patience!

**Comment by denuoweb** [comment]
I am also looking to enter a pin on a duo push. Where you suggesting that Chrome extension has this ability to test? I am on firefox, I will test the chrome if so.

![Image](https://github.com/user-attachments/assets/48863456-17c5-4ca7-bb84-2788d05fc3bd)

An error occured:

Failed to fetch https://api-46217189.duosecurity.com/push/v2/device/transactions/fb6066ae96a135d7689996cc6152a79e9903cb6466a66a?answer=approve:

Bad Request (400):
{
 "code": 40014,
 "message": "Please enter the verification code to respond to this login.",
 "stat": "FAIL"
}

![Image](https://github.com/user-attachments/assets/3ab29653-142e-472c-a8a8-8f75721982b8)

**Comment by denuoweb** [comment]
@FreshSupaSulley I tried with the chrome unpacking, I get this error: ReferenceError: urgID is not defined

Maybe it is suppose to be orgID instead of urgId?

**Comment by FreshSupaSulley** [comment]
Thank you. Stand by for a patch

**Comment by FreshSupaSulley** [comment]
@denuoweb it was a dumb fix, close to what you mentioned. Please try again and let me know what breaks.

**Comment by denuoweb** [comment]
```
@FreshSupaSulley I was now able to type in the 3 digits. When I clicked enter I got this error:
TypeError: Cannot read properties of undefined (reading 'length')

{
  "response": {
    "current_time": 1749059413,
    "transactions": [
      {
        "access_device_ble_status": "unknown",
        "akey": "****",
        "attributes": [
          [
            [
              "Organization",
              "Oregon State University"
            ],
            [
              "Integration",
              "OSU Login"
            ]
          ],
          [
            [
              "Username",
              ""****","
            ],
            [
              "IP Address",
              ""****","
            ],
            [
              "Location",
              ""****","
            ],
            [
              "Time",
              1749059401.1871352
            ]
          ]
        ],
        "block_biometric_pin_fallback": false,
        "expiration": 1749059460,
        "is_proximity_push": false,
        "is_strict_proximity_push": false,
        "phone_notification_type": "visible_notification",
        "require_lock": false,
        "require_touch_id": false,
        "step_up_code_info": {
          "num_digits": 3
        },
        "summary": "OSU Login",
        "txid": ""****",",
        "type": "Login",
        "urgid": ""****","
      }
    ]
  },
  "stat": "OK"
}
```

**Comment by FreshSupaSulley** [comment]
Made another push. Sorry about making you the guinea pig! I can't test this myself

**Comment by FreshSupaSulley** [comment]
@Eldasor not sure why it's giving you this... through some quick testing, I think I have the proper approval down?

It's currently sending a POST request like this, which is how I thought things should work:
```
https://api-blah.duosecurity.com/push/v2/device/transactions/my_urgid?step_up_code=123&step_up_code_autofilled=false&answer=approve&txId=my_urgid
```

I'm basing the flow off of this:

> [@raxod502](https://github.com/raxod502) and I did some reverse-engineering and seem to have figured out that the way to submit an approval request with verification code is to pass the code in a `step_up_code` query parameter:
> 
> [Auto-2FA/firefox/worker.js](https://github.com/FreshSupaSulley/Auto-2FA/blob/75cff6c1de8cc8833b5860ed593c144de0460472/firefox/worker.js#L190)
> 
> Line 190 in [75cff6c](/FreshSupaSulley/Auto-2FA/commit/75cff6c1de8cc8833b5860ed593c144de0460472)
> 
>  await buildRequest(info, "POST", "/push/v2/device/transactions/" + urgID, { answer: "approve" }, { txId: urgID }); 
> e.g., replace `{ answer: "approve" }` with `{ answer: "approve", step_up_code: "1234" }`
> 
> It worked for me; perhaps someone else can try and see if it works for them as well?

Any ideas?
For future reference, please include some additional logs / sanitized network requests to help me debug.

**Comment by denuoweb** [comment]
@FreshSupaSulley With the version update that was just released 10 min ago, this is the error after typing in the 3 digits, I removed some middle digits of the tx identifiers:

**Console Logs:**
ABOUT TO APPROVE A VERIFIED LOGIN!
popup.js:1090 UrgID:  9e31a9ca0fab8898dac266db8d0f5a74dd8fecca5
popup.js:1091 Transactions:  [{…}]
worker.js:257 Found transaction matching UrgID:  9e31a9ca0fab8898dac266d39a92bfab8d0f5a74dd8fecca5
worker.js:343 <pre>Unauthorized (401):<br>{
 "code": 40103,
 "message": "Invalid signature in request credentials",
 "stat": "FAIL"
}</pre>
(anonymous) @ worker.js:343
Promise.catch
buildRequest @ worker.js:342
await in buildRequest
approveTransaction @ worker.js:259
approveTransactionHandler @ worker.js:55
(anonymous) @ worker.js:34
popup.js:1110 Failed to fetch https://api-46217189.duosecurity.com/push/v2/device/transactions/9e31a9ca0fab889892bfab8d0f5a74dd8fecca5?step_up_code=822&step_up_code_autofilled=false&answer=approve&txId=9e31a9ca0fab88986aee69ec805e9a92bfab8d0f5a74dd8fecca5:<br><br><pre>Unauthorized (401):<br>{
 "code": 40103,
 "message": "Invalid signature in request credentials",
 "stat": "FAIL"
}</pre>
(anonymous) @ popup.js:1110
worker.js:329 Fetch failed loading: POST "https://api-46217189.duosecurity.com/push/v2/device/transactions/9e31a9ca0fab8898dac22bfab8d0f5a74dd8fecca5?step_up_code=822&step_up_code_autofilled=false&answer=approve&txId=9e31a9ca0fab8898dee69ec805e9a92bfab8d0f5a74dd8fecca5".
buildRequest @ worker.js:329
await in buildRequest
approveTransaction @ worker.js:259
approveTransactionHandler @ worker.js:55
(anonymous) @ worker.js:34

**Payload:**
step_up_code
986
step_up_code_autofilled
false
answer
approve
txId
8f54f7c2879b912ea8f44ebbae5a3b09a7f188b880c264672a

**Error:**
401 Unauthorized - https://api-46217189.duosecurity.com/push/v2/device/transactions/8f54f7c2879b912ea8f459bae5a3b09a7f188b880c264672a?step_up_code=986&step_up_code_autofilled=false&answer=approve&txId=8f54f7c2879b912ea824680a64ebbae5a3b09a7f188b880c264672a 

{code: 40103, message: "Invalid signature in request credentials", stat: "FAIL"}
code
: 
40103
message
: 
"Invalid signature in request credentials"
stat
: 
"FAIL"

**Comment by FreshSupaSulley** [comment]
Ok... this looks like it's supposed to be working... let's try it without `step_up_code_autofilled` attached and see what happens (just pushed!).

**Comment by FreshSupaSulley** [comment]
I'll push to prod with these changes soon. Although there's code for it, if zero-clicking breaks for verified logins, I won't know. Just drop logs / network requests if so.

**Comment by chenxiaoqino** [comment]
Worked great under Chrome. Looking forward to the release!

**Comment by FreshSupaSulley** [comment]
Thanks! It's coming soon. Wanted to give it a few days on Firefox before releasing to chrome to allow for people to test it as patching Firefox has less users + near immediate approval time, so if something goes catastrophically wrong the damage is mitigated.

**Comment by sclsj** [comment]
@FreshSupaSulley I just got access to an account that I can trigger Duo verification code consistently. However it looks like the feature is already implemented. Would sharing the account with you still be helpful?

**Comment by FreshSupaSulley** [comment]
Now that it's implemented I don't think that's necessary, also I'm concerned that's going against some sort of policy of the university (: but thank you
